src/passes/OptimizeInstructions.cpp

```diff
@@ -501,30 +501,42 @@ struct OptimizeInstructions
         }
         // math operations on a constant power of 2 right side can be optimized
         if (right->type == Type::i32) {
-          uint32_t c = right->value.geti32();
-          if (IsPowerOf2(c)) {
+          BinaryOp op;
+          int32_t c = right->value.geti32();
+          if (c >= 0 &&
+              (op = makeUnsignedBinaryOp(binary->op)) != InvalidBinary &&
```

Comment:
This really needs a comment, as it's not obvious why unsigned is better.

From the discussion it sounds like CPUs run them more quickly - interesting, is there a link for that? And that they enable further optimization. That seems reasonable, but still worth a comment.

A downside though is that for wasm2js unsigned operations will be larger (`(x | 0) > ..` vs `(x >>> 0) > ..`) but maybe that's not a strong enough reason to not do this (I'd really not want us to add an optimization flag whether we are in wasm2js mode).

Reply:
I'll add comments.
> A downside though is that for wasm2js unsigned operations will be larger ((x | 0) > .. vs (x >>> 0) > ..) 

In the past we already have similar minor regression specific to `wasm2js`. So will be great have some ability to determine compiler target inside passes (wasm or asmjs-like). It will provide more graceful optimizations which will be optimal for specific target. WDYT?

Reply:
> From the discussion it sounds like CPUs run them more quickly - interesting

Yes. During lowering to low level instruction unsigned version of division for example more simpler:
https://godbolt.org/z/bxnj3o

Reply:
https://github.com/WebAssembly/binaryen/pull/3187

Reply:
I see, thanks @MaxGraey ! Interesting about the CPU cycles there.

Yes, we may need to add a way to know if the target is wasm2js. But as I said I hope we can avoid it, that would be annoying. For now I think we are ok.